compendium/DeclarativeServices/src/ComponentContextImpl.cpp

```diff
@@ -241,34 +233,27 @@ std::shared_ptr<void> ComponentContextImpl::LocateService(
   if (!configManagerPtr) {
     throw ComponentException("Context is invalid");
   }
-  std::shared_ptr<void> service;
   auto boundServicesCacheHandle = boundServicesCache.lock();
   auto serviceMapItr = boundServicesCacheHandle->find(name);
   if (serviceMapItr != boundServicesCacheHandle->end()) {
     auto& serviceMaps = serviceMapItr->second;
     if (!serviceMaps.empty()) {
-      bool isSIPointerFoundOrMapEmpty = false;
-      service = GetInterfacePointerAndCheckError(
-        serviceMaps.at(0), type, isSIPointerFoundOrMapEmpty);
-
-      /* If a valid pointer was returned from GetInterfacePointerAndCheckError and
-         the pointer was found in the map or a nullptr was returned from
-         GetInterfacePointerAndCheckError and it was because it wasn't in the map,
-         just return service (whatever it may be).
-
-         The case where service != nullptr and isSIPointerFoundOrMapEmpty = false
-         is not possible.
-      
-         If a nullptr was returned AND it was found in the map, then something
-         went wrong during the service's construction or activation and proceed
-         with error handling.
-      */
-      bool isValid = (static_cast<bool>(service) == isSIPointerFoundOrMapEmpty);
+      ServiceInterfacePointerLookupInfo lookupInfo =
+        GetInterfacePointer(serviceMaps.at(0), type);
+
+      /* The returned service and boolean in the ServiceInterfacePointerLookupInfo struct
+       * are valid if the service pointer is null because it was not found in the map or if the
+       * service pointer is not null and it was found in the map.
+       * 
+       * If this condition is not true, then we proceed with error checking and throw
+       * a ComponentException if necessary.
+       */
+      bool isValid = (static_cast<bool>(lookupInfo.service) ==
+                      lookupInfo.wasFoundInMapOrMapEmpty);
       if (isValid) {
-        return service;
+        return lookupInfo.service;
       } else { // Checking for error condition and throwing if necessary
-        ValidateExtractedService(
-          configManagerPtr, name, type, service, isSIPointerFoundOrMapEmpty);
+        ValidateExtractedService(configManagerPtr, name, type, lookupInfo);
```

Comment:
It may be simpler to put all this logic into a single function that will either return the service object or throw.